replica/tablets.cc

```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+    {
+        _sstable_sets.reserve(sgm.storage_groups().size());
+        for (const auto& sg : sgm.storage_groups()) {
+            if (sg) {
+                auto set = sg->make_sstable_set();
+                _size += set->size();
+                _bytes_on_disk += set->bytes_on_disk();
+                _sstable_sets.emplace_back(std::move(set));
+            } else {
+                _sstable_sets.emplace_back();
+            }
+        }
+    }
+
+    static lw_shared_ptr<sstables::sstable_set> make(schema_ptr s, const storage_group_manager& sgm) {
+        return make_lw_shared<sstables::sstable_set>(std::make_unique<tablet_sstable_set>(std::move(s), sgm));
+    }
+
+    const schema_ptr& schema() const noexcept {
+        return _schema;
+    }
+
+    virtual std::unique_ptr<sstable_set_impl> clone() const override {
+        return std::make_unique<tablet_sstable_set>(*this);
+    }
+
+    virtual std::vector<sstables::shared_sstable> select(const dht::partition_range& range = query::full_partition_range) const override;
+    virtual lw_shared_ptr<const sstable_list> all() const override;
+    virtual stop_iteration for_each_sstable_until(std::function<stop_iteration(const sstables::shared_sstable&)> func) const override;
+    virtual future<stop_iteration> for_each_sstable_gently_until(std::function<future<stop_iteration>(const sstables::shared_sstable&)> func) const override;
+    virtual bool insert(sstables::shared_sstable sst) override;
+    virtual bool erase(sstables::shared_sstable sst) override;
+    virtual size_t size() const noexcept override {
+        return _size;
+    }
+    virtual uint64_t bytes_on_disk() const noexcept override {
+        return _bytes_on_disk;
+    }
+    virtual selector_and_schema_t make_incremental_selector() const override;
+
+    virtual flat_mutation_reader_v2 create_single_key_sstable_reader(
+            replica::column_family*,
+            schema_ptr,
+            reader_permit,
+            utils::estimated_histogram&,
+            const dht::partition_range&,
+            const query::partition_slice&,
+            tracing::trace_state_ptr,
+            streamed_mutation::forwarding,
+            mutation_reader::forwarding,
+            const sstables::sstable_predicate&) const override;
+
+private:
+    size_t group_of(const dht::token& t) const noexcept {
+        return dht::compaction_group_of(_log2, t);
+    }
+    dht::token first_token_of(size_t idx) const noexcept {
+        if (!idx) {
+            return dht::minimum_token();
+        }
+        if (idx >= _sstable_sets.size()) {
+            return dht::maximum_token();
+        }
+        return dht::next_token(dht::last_token_of_compaction_group(_log2, idx - 1));
+    }
+    dht::token last_token_of(size_t idx) const noexcept {
+        if (idx >= _sstable_sets.size()) {
+            return dht::maximum_token();
+        }
+        return dht::last_token_of_compaction_group(_log2, idx);
+    }
```

Comment:
This works today, but it's assuming a particular implementation of tablet map. so knowledge duplication.

It seems you need a snapshot of tablet_map. Why don't you add tablet_map::token_whatever which stores only log2_tablets and provides all these token-related functions? tablet_map::token_whatever gets reused by tablet_map itself. Then you can call tablet_map::get_token_whatever() and store it here, which will represent the tablet map in a particular point in time.

Reply:
sounds like a good idea.
it may help convert to uneven tablet ranges if and when we get there

Reply:
that said, maybe we should do it as a follow-up since #16876 is P0 and I don't want to delay this series indefinitely (I'm sure it can be improved upon further and further)